processor/spanmetricsprocessor/processor.go

```diff
@@ -322,54 +354,82 @@ func (p *processorImp) aggregateMetricsForSpan(serviceName string, span pdata.Sp
 	// Binary search to find the latencyInMilliseconds bucket index.
 	index := sort.SearchFloat64s(p.latencyBounds, latencyInMilliseconds)
 
-	key := buildKey(serviceName, span, p.dimensions, resourceAttr)
+	mKey := p.buildMetricKey(span, resourceAttr)
+	resourceAttrKey := p.buildResourceAttrKey(serviceName, resourceAttr)
+	p.resourceAttrList[resourceAttrKey] = true
 
 	p.lock.Lock()
-	p.cache(serviceName, span, key, resourceAttr)
-	p.updateCallMetrics(key)
-	p.updateLatencyMetrics(key, latencyInMilliseconds, index)
-	p.updateLatencyExemplars(key, latencyInMilliseconds, span.TraceID())
+	p.cacheMetricKey(span, mKey, resourceAttr)
+	p.cacheResourceAttrKey(serviceName, resourceAttr, resourceAttrKey)
+	p.updateCallMetrics(resourceAttrKey, mKey)
+	p.updateLatencyMetrics(resourceAttrKey, mKey, latencyInMilliseconds, index)
+	p.updateLatencyExemplars(resourceAttrKey, mKey, latencyInMilliseconds, span.TraceID())
 	p.lock.Unlock()
 }
 
 // updateCallMetrics increments the call count for the given metric key.
-func (p *processorImp) updateCallMetrics(key metricKey) {
-	p.callSum[key]++
+func (p *processorImp) updateCallMetrics(resourceAttrKey resourceKey, mKey metricKey) {
+	if _, ok := p.callSum[resourceAttrKey]; ok {
```

Comment:
Potential for lock as well.

Reply:
The lock is held in the code that calls this function so it should be fine right?

Reply:
I worry that someone could mistake these functions safe for concurrent access and trip up.

From what I understand, it is depends on the caller to make sure this happens?

It seems okay since the method is not public, however, it does raise a question if this should be a seperate function or just one big function?

Reply:
`ConsumeTraces()` should be concurrent-safe and that's it. IMO it's kind of impractical to expect every function here to be safe, and it's reasonable to expect the caller to take concurrency into account. That is how the component is designed before we touched it anyway.

Reply:
I've put the change into this commit: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6717/commits/fd42b768ac6bd6dd1a7b541a8f9ed621002c0f48